admin/provisioner/clickhousestatic/provisioner.go

```diff
@@ -69,7 +84,48 @@ func New(specJSON []byte, _ database.DB, logger *zap.Logger) (provisioner.Provis
 		spec.DSN = dsn
 	}
 
-	opts, err := clickhouse.ParseDSN(spec.DSN)
+	if spec.ReadDSNPath != "" {
+		dsn, err := os.ReadFile(spec.ReadDSNPath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read read DSN file: %w", err)
+		}
+		spec.ReadDSN = strings.TrimSpace(string(dsn))
+	}
+
+	if spec.ReadDSNEnv != "" && spec.ReadDSN == "" && spec.ReadDSNPath == "" {
+		dsn := os.Getenv(spec.ReadDSNEnv)
+		if dsn == "" {
+			return nil, fmt.Errorf("environment variable %s is not set or empty", spec.ReadDSNEnv)
+		}
+		spec.ReadDSN = dsn
+	}
+
+	if spec.WriteDSNPath != "" {
+		dsn, err := os.ReadFile(spec.WriteDSNPath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read write DSN file: %w", err)
+		}
+		spec.WriteDSN = strings.TrimSpace(string(dsn))
+	}
+
+	if spec.WriteDSNEnv != "" && spec.WriteDSN == "" && spec.WriteDSNPath == "" {
+		dsn := os.Getenv(spec.WriteDSNEnv)
+		if dsn == "" {
+			return nil, fmt.Errorf("environment variable %s is not set or empty", spec.WriteDSNEnv)
+		}
+		spec.WriteDSN = dsn
+	}
+
+	provDSN := spec.DSN
+	if provDSN == "" && spec.WriteDSN != "" {
+		provDSN = spec.WriteDSN
+	}
+
+	if provDSN == "" {
+		return nil, fmt.Errorf("no DSN available - must specify either 'dsn' or 'write_dsn'")
+	}
```

Comment:
Could the logic be simplified if you just have `DSN` (required) and `WriteDSN` (optional) options? Then it uses `WriteDSN` for writes if set, otherwise always falls back to `DSN`.

It avoids some of these edge cases where `ReadDSN` and `WriteDSN` combined are mutually exclusive of `DSN`.